## Условие 1: 1 задача
Скачать файл в уроке.
Считать данные с помощью pandas.
Вывести на экран первые 5 строк.
Посмотреть на описание признаков и на их содержание.

In [1]:
import pandas as pd


In [2]:
df = pd.read_csv('/content/kc_final (1).csv')
df

Unnamed: 0          id        date     price  bedrooms  bathrooms  \
0               0  7129300520  2014-10-13  221900.0         3       1.00   
1               1  6414100192  2014-12-09  538000.0         3       2.25   
2               2  5631500400  2015-02-25  180000.0         2       1.00   
3               3  2487200875  2014-12-09  604000.0         4       3.00   
4               4  1954400510  2015-02-18  510000.0         3       2.00   
...           ...         ...         ...       ...       ...        ...   
21608       21608   263000018  2014-05-21  360000.0         3       2.50   
21609       21609  6600060120  2015-02-23  400000.0         4       2.50   
21610       21610  1523300141  2014-06-23  402101.0         2       0.75   
21611       21611   291310100  2015-01-16  400000.0         3       2.50   
21612       21612  1523300157  2014-10-15  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0             1180      5650     1.0           0  ...      7        1180   
1             2570      7242     2.0           0  ...      7        2170   
2              770     10000     1.0           0  ...      6         770   
3             1960      5000     1.0           0  ...      7        1050   
4             1680      8080     1.0           0  ...      8        1680   
...            ...       ...     ...         ...  ...    ...         ...   
21608         1530      1131     3.0           0  ...      8        1530   
21609         2310      5813     2.0           0  ...      8        2310   
21610         1020      1350     2.0           0  ...      7        1020   
21611         1600      2388     2.0           0  ...      8        1600   
21612         1020      1076     2.0           0  ...      7        1020   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0                  0      1955             0    98178  47.5112 -122.257   
1                400      1951          1991    98125  47.7210 -122.319   
2                  0      1933             0    98028  47.7379 -122.233   
3                910      1965             0    98136  47.5208 -122.393   
4                  0      1987             0    98074  47.6168 -122.045   
...              ...       ...           ...      ...      ...      ...   
21608              0      2009             0    98103  47.6993 -122.346   
21609              0      2014             0    98146  47.5107 -122.362   
21610              0      2009             0    98144  47.5944 -122.299   
21611              0      2004             0    98027  47.5345 -122.069   
21612              0      2008             0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21608           1530        1509  
21609           1830        7200  
21610           1020        2007  
21611           1410        1287  
21612           1020        1357  

[21613 rows x 22 columns]

In [3]:
# Первые пять строк таблицы df
df.head()

Unnamed: 0          id        date     price  bedrooms  bathrooms  \
0           0  7129300520  2014-10-13  221900.0         3       1.00   
1           1  6414100192  2014-12-09  538000.0         3       2.25   
2           2  5631500400  2015-02-25  180000.0         2       1.00   
3           3  2487200875  2014-12-09  604000.0         4       3.00   
4           4  1954400510  2015-02-18  510000.0         3       2.00   

   sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
0         1180      5650     1.0           0  ...      7        1180   
1         2570      7242     2.0           0  ...      7        2170   
2          770     10000     1.0           0  ...      6         770   
3         1960      5000     1.0           0  ...      7        1050   
4         1680      8080     1.0           0  ...      8        1680   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 22 columns]

В дата фрейме представлены данные о продающихся домах и приведены некоторые описания. Среди них: идентификационный номер дома, цена дома, количество спален, количество ванных комнат, жилая площадь, общая площадь, количество этажей, оценка дома, площадь цокольного этажа, год постройки, год капитального ремонта.

##Условие 2: 2 задача
Проведите первичный анализ данных.
Изучите типы данных.
Найдите количество пропущенных ячеек в данных.
Посчитайте основные статистики по всем признакам и поизучайте их.
Пишите выводы.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     21613 non-null  int64  
 1   id             21613 non-null  int64  
 2   date           21613 non-null  object 
 3   price          21613 non-null  float64
 4   bedrooms       21613 non-null  int64  
 5   bathrooms      21613 non-null  float64
 6   sqft_living    21613 non-null  int64  
 7   sqft_lot       21613 non-null  int64  
 8   floors         21613 non-null  float64
 9   waterfront     21613 non-null  int64  
 10  view           21613 non-null  int64  
 11  condition      21613 non-null  int64  
 12  grade          21613 non-null  int64  
 13  sqft_above     21613 non-null  int64  
 14  sqft_basement  21613 non-null  int64  
 15  yr_built       21613 non-null  int64  
 16  yr_renovated   21613 non-null  int64  
 17  zipcode        21613 non-null  int64  
 18  lat   

In [5]:
df.shape

(21613, 22)

В дата фрейме в основном представлены числовые данные в формате целочисленные и с десятичной запятой. Кроме даты, которая имеет строковый формат. Всего 22 столбца Все (21613) строки заполнены данными, пропусков данных нет.

In [6]:
# Основные статистики
df.describe()

Unnamed: 0            id         price      bedrooms     bathrooms  \
count  21613.00000  2.161300e+04  2.161300e+04  21613.000000  21613.000000   
mean   10806.00000  4.580302e+09  5.400881e+05      3.370842      2.114757   
std     6239.28002  2.876566e+09  3.671272e+05      0.930062      0.770163   
min        0.00000  1.000102e+06  7.500000e+04      0.000000      0.000000   
25%     5403.00000  2.123049e+09  3.219500e+05      3.000000      1.750000   
50%    10806.00000  3.904930e+09  4.500000e+05      3.000000      2.250000   
75%    16209.00000  7.308900e+09  6.450000e+05      4.000000      2.500000   
max    21612.00000  9.900000e+09  7.700000e+06     33.000000      8.000000   

        sqft_living      sqft_lot        floors    waterfront          view  \
count  21613.000000  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean    2079.899736  1.510697e+04      1.494309      0.007542      0.234303   
std      918.440897  4.142051e+04      0.539989      0.086517      0.766318   
min      290.000000  5.200000e+02      1.000000      0.000000      0.000000   
25%     1427.000000  5.040000e+03      1.000000      0.000000      0.000000   
50%     1910.000000  7.618000e+03      1.500000      0.000000      0.000000   
75%     2550.000000  1.068800e+04      2.000000      0.000000      0.000000   
max    13540.000000  1.651359e+06      3.500000      1.000000      4.000000   

       ...         grade    sqft_above  sqft_basement      yr_built  \
count  ...  21613.000000  21613.000000   21613.000000  21613.000000   
mean   ...      7.656873   1788.390691     291.509045   1971.005136   
std    ...      1.175459    828.090978     442.575043     29.373411   
min    ...      1.000000    290.000000       0.000000   1900.000000   
25%    ...      7.000000   1190.000000       0.000000   1951.000000   
50%    ...      7.000000   1560.000000       0.000000   1975.000000   
75%    ...      8.000000   2210.000000     560.000000   1997.000000   
max    ...     13.000000   9410.000000    4820.000000   2015.000000   

       yr_renovated       zipcode           lat          long  sqft_living15  \
count  21613.000000  21613.000000  21613.000000  21613.000000   21613.000000   
mean      84.402258  98077.939805     47.560053   -122.213896    1986.552492   
std      401.679240     53.505026      0.138564      0.140828     685.391304   
min        0.000000  98001.000000     47.155900   -122.519000     399.000000   
25%        0.000000  98033.000000     47.471000   -122.328000    1490.000000   
50%        0.000000  98065.000000     47.571800   -122.230000    1840.000000   
75%        0.000000  98118.000000     47.678000   -122.125000    2360.000000   
max     2015.000000  98199.000000     47.777600   -121.315000    6210.000000   

          sqft_lot15  
count   21613.000000  
mean    12768.455652  
std     27304.179631  
min       651.000000  
25%      5100.000000  
50%      7620.000000  
75%     10083.000000  
max    871200.000000  

[8 rows x 21 columns]

**Выводы.** Самый дешевый дом стоит 75 000, самый дорогой дом стоит 7 700 000. Средняя цена дома составляет 540 000, а стандартное отклоненик составляет +-367 000. Минимальное кол-во спален 0, макс-ое кол-во спален 33, среднее кол-во спален 3,371. Минимальное кол-во ванных комнат 0, макс-ое кол-во ванных комнат  8, среднее кол-во ванных комнат 2,114. Самая маленькая жилая площадь 290sqft, самая большая жилая площадь 13540sqft, средняя жилая площадь 2079,899sqft. Самая маленькая общая площадь 520sqft, самая большая общая площадь 1,65 million sqft, средняя общая площадь 15 100sqft. Средняя этажность 1,49. Самый маленький рейтинг 1, самый большой 13. Самый старый дом 1900го года постройки, у него не было капитального ремонта. Самый молодой дом - 2015 года постройки и в этом же году он претерпел конструктивные изменения.

## Условие 3: 3 задача
Ответьте на несколько вопросов.

### 3.1
 В каком диапазоне изменяются стоимости недвижимости?

In [7]:

# Минимальное значение:
df['price'].min()

75000.0

In [8]:
# Максимальное значение:
df['price'].max()

7700000.0

###3.2
Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [9]:
# Средняя жилая площадь
df['sqft_living'].mean()

2079.8997362698374

In [10]:
# Средняя общая площадь
df['sqft_lot'].mean()

15106.967565816869

In [11]:
# Доля в среднем жилой площади от общей площади
part = (df['sqft_living'].mean()) / (df['sqft_lot'].mean())
part

0.1376781757959227

###3.3
 Как много домов с разными этажами в данных?

In [12]:
df['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

###3.4
Насколько хорошие состояния у домов в данных?

In [13]:
df['yr_built'].value_counts()

2014    559
2006    454
2005    450
2004    433
2003    422
       ... 
1933     30
1901     29
1902     27
1935     24
1934     21
Name: yr_built, Length: 116, dtype: int64

In [14]:
df['yr_renovated'].value_counts()

0       20699
2014       91
2013       37
2003       36
2005       35
        ...  
1951        1
1959        1
1948        1
1954        1
1944        1
Name: yr_renovated, Length: 70, dtype: int64

**Вывод:** Из 21613 домов 20699 домов никогда не ремонтировались, дата постройки начиналась с 1934 года. Отсюда можно сделать вывод, что у большинства домов состояние неудовлетворительное.

###3.5
Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [15]:
# Год пострйки самого первого дома в дата фрейме
df['yr_built'].min()

1900

In [16]:
# Год пострйки самого последнего дома в дата фрейме
df['yr_built'].max()

2015

## Условие 4: 4 задача
Ответьте на несколько вопросов

###4.1
Сколько в среднем стоят дома, у которых 2 спальни?

In [17]:
df[df['bedrooms'] == 2]['price'].mean()

401372.681884058

###4.2
Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [18]:
df[(df['price'] > 600000)]['sqft_lot'].mean()

20442.524776214832

###4.3
Как много домов коснулся ремонт?

In [19]:
df[(df['yr_renovated'] != 0)]

Unnamed: 0          id        date      price  bedrooms  bathrooms  \
1               1  6414100192  2014-12-09   538000.0         3       2.25   
35             35  9547205180  2014-06-13   696000.0         3       2.50   
95             95  1483300570  2014-09-08   905000.0         4       2.50   
103           103  2450000295  2014-10-07  1088000.0         3       2.50   
115           115  3626039325  2014-11-21   740500.0         3       3.50   
...           ...         ...         ...        ...       ...        ...   
19622       19622  7351200295  2015-01-14  1150000.0         3       1.75   
20057       20057   126039256  2014-09-04   434900.0         3       2.00   
20444       20444  4305600360  2015-02-25   500012.0         4       2.50   
20447       20447  3319500628  2015-02-12   356999.0         3       1.50   
20962       20962  1278000210  2015-03-11   110000.0         2       1.00   

       sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
1             2570      7242     2.0           0  ...      7        2170   
35            2300      3060     1.5           0  ...      8        1510   
95            3300     10250     1.0           0  ...      7        2390   
103           2920      8113     2.0           0  ...      8        2920   
115           4380      6350     2.0           0  ...      8        2780   
...            ...       ...     ...         ...  ...    ...         ...   
19622         1760      6788     2.0           1  ...      7        1760   
20057         1520      5040     2.0           0  ...      7        1520   
20444         2400      9612     1.0           0  ...      8        1230   
20447         1010      1546     2.0           0  ...      8        1010   
20962          828      4524     1.0           0  ...      6         828   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
1                400      1951          1991    98125  47.7210 -122.319   
35               790      1930          2002    98115  47.6827 -122.310   
95               910      1946          1991    98040  47.5873 -122.249   
103                0      1950          2010    98004  47.5814 -122.196   
115             1600      1900          1999    98117  47.6981 -122.368   
...              ...       ...           ...      ...      ...      ...   
19622              0      1940          1960    98125  47.7336 -122.284   
20057              0      1977          2006    98177  47.7770 -122.362   
20444           1170      1962          2009    98059  47.4799 -122.127   
20447              0      1971          2014    98144  47.5998 -122.311   
20962              0      1968          2007    98001  47.2655 -122.244   

       sqft_living15  sqft_lot15  
1               1690        7639  
35              1590        3264  
95              1950        6045  
103             2370        8113  
115             1830        6350  
...              ...         ...  
19622           1630        7588  
20057           1860        8710  
20444           2430        5539  
20447           1010        1517  
20962            828        5402  

[914 rows x 22 columns]

**Вывод:** Кол-во домов, которых коснулся ремонт равен 914.

###4.4
Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [20]:
# в среднем стоимость домов с оценкой grade выше 10
df[df['grade'] > 10]['price'].mean()

1678635.1175298805

In [21]:
# в среднем стоимость домов с оценкой grade меньше 4
df[df['grade'] < 4]['price'].mean()

189750.0

In [22]:
# в среднем стоимость домов с оценкой grade выше 10 отличается от стоимости домов с оценкой grade меньше 4 на:
diff = df[df['grade'] > 10]['price'].mean() - df[df['grade'] < 4]['price'].mean()
print(format(diff, '.2f'))

1488885.12


##Условие 5: 5 задача

###5.1
Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [28]:
df[(df['waterfront'] == 1) & (df['bathrooms'] >= 3) & (df['sqft_basement'] > 0)]

Unnamed: 0          id        date      price  bedrooms  bathrooms  \
300           300  3225069065  2014-06-24  3075000.0         4       5.00   
457           457  2122039094  2014-11-26   705000.0         3       3.00   
1164         1164  1247600105  2014-10-20  5110800.0         5       5.25   
1261         1261  7636800041  2014-06-25   995000.0         3       4.50   
1315         1315  7558700030  2015-04-13  5300000.0         6       6.00   
1361         1361  1925069082  2015-05-11  2200000.0         5       4.25   
2626         2626  7738500731  2014-08-15  4500000.0         5       5.50   
2864         2864  4114601570  2014-11-18  3600000.0         3       3.25   
3914         3914  9808700762  2014-06-11  7062500.0         5       4.50   
3934         3934  1269200229  2014-07-23  1379900.0         3       3.25   
5760         5760   623059016  2014-07-17  1100000.0         4       3.25   
6233         6233  2024069008  2014-06-19  2200000.0         5       4.75   
6508         6508  4217402115  2015-04-21  3650000.0         6       4.75   
7989         7989  9362000040  2014-06-23  3400000.0         3       4.50   
8092         8092  1924059029  2014-06-17  4668000.0         5       6.75   
8163         8163  1724069079  2015-03-19  1452000.0         2       3.25   
8784         8784  3343302110  2015-03-06  1800000.0         3       3.00   
10263       10263   625059036  2014-08-13  2700000.0         5       4.00   
11257       11257  3761100045  2014-06-18  3000000.0         4       4.25   
11402       11402  3426049284  2014-08-19  2300000.0         4       3.25   
11424       11424  5119400075  2014-06-20   950000.0         3       3.25   
11446       11446   624069035  2014-12-09  2750000.0         4       4.00   
11844       11844  2623039082  2015-02-18   770000.0         3       3.50   
12459       12459  4114601580  2014-07-24  1900000.0         6       4.00   
12935       12935  3222049044  2014-06-12   835000.0         3       3.00   
13710       13710  2923500230  2014-12-16  2600000.0         4       4.50   
13967       13967  3126059023  2015-03-03  3395000.0         4       3.50   
14083       14083  3625059043  2014-09-04  3300000.0         5       4.75   
14619       14619  7936500109  2014-07-25  2230000.0         3       3.00   
14822       14822  3623500408  2015-03-30  2600000.0         3       3.00   
15482       15482   624069108  2014-08-12  3200000.0         4       3.25   
15668       15668  4030100290  2014-10-01  1680000.0         5       3.50   
16817       16817   825059178  2014-09-23  2574000.0         4       3.75   
17767       17767  7352200450  2015-01-15  2050000.0         4       3.25   
18199       18199  3625059152  2014-12-30  3300000.0         3       3.25   
18455       18455  8043700300  2014-06-08  2700000.0         4       3.25   
18556       18556  6613000930  2014-09-02  2950000.0         4       3.25   
20325       20325   518500480  2014-08-11  3000000.0         3       3.50   
20767       20767  8043700105  2015-04-17  2300000.0         4       4.00   
21201       21201   518500460  2014-10-08  2230000.0         3       3.50   
21576       21576  9253900271  2015-01-07  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  ...  grade  sqft_above  \
300           4550     18641     1.0           1  ...     10        2600   
457           1970     20978     2.0           1  ...      9        1770   
1164          8010     45517     2.0           1  ...     12        5990   
1261          4380     47044     2.0           1  ...      9        3720   
1315          7390     24829     2.0           1  ...     12        5000   
1361          4640     22703     2.0           1  ...      8        2860   
2626          6640     40014     2.0           1  ...     12        6350   
2864          5020     12431     2.0           1  ...     10        3420   
3914         10040     37325     2.0           1  ...     11        7680   
3934        

**Вывод:** У клиента есть 41 вариант домов.

###5.2
Выберите дом клиенту.
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [24]:
df['view'].value_counts()

0    19489
2      963
3      510
1      332
4      319
Name: view, dtype: int64

In [26]:
df[(df['view'] == 4) | (df['waterfront'] == 1) & (df['yr_renovated'] != 0) & (df['yr_built'] >= 1980)]['price'].min()

252000.0

In [27]:
df[(df['view'] == 4) | (df['waterfront'] == 1) & (df['yr_renovated'] != 0) & (df['yr_built'] >= 1980)]['price'].max()

6885000.0

**Вывод:** Дома будут в ценовом диапазоне: min = 252 000, max = 6 885 000.

###5.3
Выберите дом клиенту. Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [31]:
df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] <= 150000)][['yr_built', 'yr_renovated']]
# df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] <= 150000)]

yr_built  yr_renovated
2589       1922          2003
10105      1983             0
11255      1978             0
15610      1981             0
18332      1908             0
18939      1969             0
18988      1966             0

**Вывод:** Всего семь вариантов. Самый "старый дом" - 1908 года постройки, самый "молодой" дом - 1983 года постройки. Причем из семи домов, только один дом, 1922 года постройки, был отремонтирован в 2003 году. Поэтому можно рассматривать три варианта: упомянутый дом 1922 года постройки и 2003 года ремонта, а также дома постройки 1981 и 1983 годов, ремонт которых не проводился.